mdbook-tera-backend/src/tera_renderer/renderer.rs

```diff
@@ -0,0 +1,140 @@
+    /// by considering all the output HTML files as input to the Tera template.
+    /// It overwrites the preexisting files with their Tera-rendered version.
+    pub(crate) fn render_book(&mut self) -> Result<()> {
+        let dest_dir = self
+            .ctx
+            .destination
+            .parent()
+            .unwrap()
+            .join("html")
+            .to_owned();
+        if !dest_dir.is_dir() {
+            return Err(anyhow!(
+                "{dest_dir:?} is not a directory. Please make sure the HTML renderer is enabled."
+            ));
+        }
+        self.render_book_directory(&dest_dir)
+    }
+
+    /// Create the `get_context` function for the `Tera` template, a helper that allows retrieving values from `ctx`.
+    fn create_get_context_function(&self) -> impl tera::Function {
+        let ctx_rc = Arc::clone(&self.ctx);
+        move |args: &HashMap<String, serde_json::value::Value>| -> tera::Result<tera::Value> {
+            let key = args
+                .get("key")
+                .ok_or_else(|| tera::Error::from(format!("No key argument provided")))?
+                .as_str()
+                .ok_or_else(|| {
+                    tera::Error::from(format!("Key has invalid type, expected string"))
+                })?;
+            let value = ctx_rc
+                .config
+                .get(key)
+                .ok_or_else(|| tera::Error::from(format!("Could not find key {key} in config")))?;
+            let value = to_value(value)?;
+            Ok(value)
+        }
+    }
+
+    /// Render the book directory located at `path` recursively.
+    fn render_book_directory(&mut self, path: &Path) -> Result<()> {
+        for entry in path.read_dir()? {
+            let entry = entry?;
+            let path = entry.path();
+            if path.is_dir() {
+                self.render_book_directory(&path)?;
+            } else {
+                self.process_file(&path)?;
+            }
+        }
+        Ok(())
+    }
+
+    /// Reads the file at `path` and renders it.
+    fn process_file(&mut self, path: &Path) -> Result<()> {
+        if path.extension().unwrap_or_default() != "html" {
+            return Ok(());
+        }
+        let file_content = std::fs::read_to_string(path)?;
+        let output = self.render_file_content(&file_content, path)?;
+        let mut output_file = fs::File::create(path)?;
+        output_file.write_all(output.as_bytes())?;
+        Ok(())
+    }
+
+    /// Creates the rendering context to be passed to the templates.
+    ///
+    /// # Arguments
+    ///
+    /// `path`: The path to the file that will be added as extra context to the renderer.
+    fn create_context(&mut self, path: &Path) -> Result<tera::Context> {
+        let mut context = tera::Context::new();
+        context.insert("path", path);
+        context.insert("ctx", &serde_json::to_value(&*self.ctx)?);
+        context.insert("book_dir", &self.ctx.destination.parent().unwrap());
```

Comment:
This can be computed from `ctx` so it looks unnecessary?

Reply:
That's annoying to compute from `ctx` in tera. I think it's a good addition

Reply:
It can be computed with a small pipeline:

```jinja
{{ ctx.destination | split(pat="/") | slice(end=-1) | join(sep="/") }}
```